### PR TITLE
WIP: Implement configurable LSTM

### PR DESCRIPTION
**Pull request by nemerchiedde**

### CODE REVIEW

File: hls4ml/converters/keras_to_hls.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 import math
 
 from hls4ml.model import HLSModel
+#from hls4ml.writer.quartus_writer import write_activation_lstm
```
Comment (etfortin) [inline hls4ml/converters/keras_to_hls.py:8]: Remove  comment

File: hls4ml/converters/keras_to_hls.py [hunk 2]
```diff
@@ -114,31 +116,82 @@ def parse_default_keras_layer(keras_layer, input_names):
...
+              second_split_ractv[0] = layer_rec_activation
+              res = "\tnnet::" + "<data_T".join(second_split_ractv)
+              print(res, 'qual é a linha2')
+          
+       else:
+           res = my_lines[i]
+       my_lines[i] = res
+       i += 1    
+         
+    with open("../hls4ml/templates/quartus/firmware/nnet_utils/lstm_cell.h","w") as myfile:
```
Comment (etfortin) [inline hls4ml/converters/keras_to_hls.py]: Change to modify output files

File: hls4ml/converters/keras_to_hls.py [hunk 3]
```diff
@@ -114,31 +116,82 @@ def parse_default_keras_layer(keras_layer, input_names):
...
     if 'epsilon' in keras_layer['config']:
         layer['epsilon'] = keras_layer['config']['epsilon']
 
     return layer
-
-
-def keras_to_hls(config):
+"""
+def modification_activation_lstm(layer_rec_activation,layer_activation):
+    with open("../hls4ml/templates/quartus/firmware/nnet_utils/lstm_cell.h","r") as myfile:
```
Comment (etfortin) [inline hls4ml/converters/keras_to_hls.py]: Change to modify output files

File: hls4ml/converters/keras_to_hls.py [hunk 4]
```diff
@@ -114,31 +116,82 @@ def parse_default_keras_layer(keras_layer, input_names):
...
         reader = KerasModelReader(config['KerasModel'])
-    elif 'KerasJson' in config:
+    elif 'KerasJson' in config:                                              ##Usamos este
         # Extract model architecture from json
         with open(config['KerasJson']) as json_file:
             model_arch = json.load(json_file)
         reader = KerasFileReader(config)
+        #print("reader",reader)
     elif 'KerasH5' in config:
+        print("terceiro")
```
Comment (etfortin) [inline hls4ml/converters/keras_to_hls.py]: Remove portugese comments

File: hls4ml/converters/keras_to_hls.py [hunk 5]
```diff
@@ -168,19 +222,25 @@ def keras_to_hls(config):
     output_layers = None
 
     layer_config = None
-    if model_arch['class_name'] == 'Sequential':
+    layer_activation = []
+
+    if model_arch['class_name'] == 'Sequential':  #model_arch contem o json
```
Comment (etfortin) [inline hls4ml/converters/keras_to_hls.py]: remove portugese comment

File: hls4ml/converters/keras_to_hls.py [hunk 6]
```diff
@@ -198,14 +258,18 @@ def keras_to_hls(config):
     for keras_layer in layer_config:
         if 'batch_input_shape' in keras_layer['config']:
             input_shapes = [keras_layer['config']['batch_input_shape']]
+            #print('input_shapes',input_shapes, "keras_layer['config']",keras_layer['config'])
```
Comment (etfortin) [inline hls4ml/converters/keras_to_hls.py]: Why modifying this part ????

File: hls4ml/converters/keras_to_hls.py [hunk 7]
```diff
@@ -249,16 +320,157 @@ def keras_to_hls(config):
...
+            #print('nemer chiedde')
+            #print('layer list 1', layer_list[1]['activation']) #recurrent_activation
+            
+            layer_activation.append(layer_list[1]['activation'])
+            layer_activation.append(layer_list[2]['activation'])
+            #print(layer_activation)  
+            #print (layer_rec_activation)
+            activation_type.append(layer_list[1]['activation'])
+            activation_type.append(layer_rec_activation)
+
```
Comment (etfortin) [inline hls4ml/converters/keras_to_hls.py]: Cleaning comment needed and why modification

File: hls4ml/converters/keras_to_hls.py (hunk 6 above)
Comment (nemerchiedde) [reply hls4ml/converters/keras_to_hls.py]: This is just the print. I used it to check where i could get the layer_activation from JSON.

File: hls4ml/converters/keras_to_hls.py (hunk 7 above)
Comment (nemerchiedde) [reply hls4ml/converters/keras_to_hls.py]: New merge doesn't has the comment. I modified  this part to get the activation and rec_activation

File: hls4ml/converters/keras_to_hls.py (hunk 5 above)
Comment (nemerchiedde) [reply hls4ml/converters/keras_to_hls.py]: New merge doesn't has the comment.

File: hls4ml/converters/keras_to_hls.py (hunk 4 above)
Comment (nemerchiedde) [reply hls4ml/converters/keras_to_hls.py]: New merge doesn't has the comment.

File: hls4ml/converters/keras_to_hls.py (hunk 3 above)
Comment (nemerchiedde) [reply hls4ml/converters/keras_to_hls.py]: New merge already fixed this problem.

File: hls4ml/converters/keras_to_hls.py (hunk 2 above)
Comment (nemerchiedde) [reply hls4ml/converters/keras_to_hls.py]: New merge already fixed this problem.

File: hls4ml/converters/keras_to_hls.py (hunk 1 above)
Comment (nemerchiedde) [reply hls4ml/converters/keras_to_hls.py:8]: New merge already done

File: hls4ml/writer/quartus_writer.py [hunk 8]
```diff
@@ -65,14 +65,15 @@ def print_array_to_cpp(self, var, layer, odir):
         h_file.close()
 
     def write_project_dir(self, model):
+        print("FOI?")
```
Comment (etfortin) [inline hls4ml/writer/quartus_writer.py]: Clean prints

File: hls4ml/writer/quartus_writer.py [hunk 9]
```diff
@@ -65,14 +65,15 @@ def print_array_to_cpp(self, var, layer, odir):
...
         if not os.path.isdir("{}/firmware/weights".format(model.config.get_output_dir())):
             os.makedirs("{}/firmware/weights".format(model.config.get_output_dir()))
 
     def write_project_cpp(self, model):
         ###################
         ## myproject.cpp
         ###################
-
+        print("FOI?")
```
Comment (etfortin) [inline hls4ml/writer/quartus_writer.py]: clean print

File: hls4ml/writer/quartus_writer.py [hunk 10]
```diff
@@ -130,8 +131,8 @@ def write_project_cpp(self, model):
                             name = var.definition_cpp_name()
                             newline += '    ' + 'hls_register outputdat ' + name + ';\n'
                             var.name += '.data'
-                    if layer.get_attr('activation') == 'tanh':
-                        layer.set_attr('activation') == 'dense_tanh'
+                 #   if layer.get_attr('activation') == 'tanh':
+                 #       layer.set_attr('activation') == 'dense_tanh'
```
Comment (etfortin) [inline hls4ml/writer/quartus_writer.py]: cleaning

File: hls4ml/writer/quartus_writer.py [hunk 11]
```diff
@@ -814,6 +875,9 @@ def write_hls(self, model):
         self.write_bridge(model)
         self.write_build_script(model)
         self.write_nnet_utils(model)
+        self.write_activation_lstm(model)
+        #self.write_input_dimention(model)
```
Comment (etfortin) [inline hls4ml/writer/quartus_writer.py]: cleaning

File: hls4ml/writer/quartus_writer.py [hunk 12]
```diff
@@ -814,6 +875,9 @@ def write_hls(self, model):
         self.write_bridge(model)
         self.write_build_script(model)
         self.write_nnet_utils(model)
+        self.write_activation_lstm(model)
```
Comment (etfortin) [inline hls4ml/writer/quartus_writer.py:933]: It is needed to have a condition, if the activation is not lstm we don't have to do this function

File: hls4ml/converters/keras_to_hls.py (hunk 4 above)
Comment (etfortin) [reply hls4ml/converters/keras_to_hls.py]: In  the files pushed in your branch it had it

File: hls4ml/converters/keras_to_hls.py [hunk 13]
```diff
@@ -168,26 +172,32 @@ def keras_to_hls(config):
     output_layers = None
 
     layer_config = None
-    if model_arch['class_name'] == 'Sequential':
+    layer_activation = []
+
+    if model_arch['class_name'] == 'Sequential':  #model_arch contem o json
```
Comment (etfortin) [inline hls4ml/converters/keras_to_hls.py]: Portugese comment

File: hls4ml/converters/keras_to_hls.py [hunk 14]
```diff
@@ -197,17 +207,24 @@ def keras_to_hls(config):
     print('Topology:')
     for keras_layer in layer_config:
         if 'batch_input_shape' in keras_layer['config']:
-            input_shapes = [keras_layer['config']['batch_input_shape']]
+            input_shapes = [keras_layer['config']['batch_input_shape']] 
+            print('input_shapes',input_shapes, "keras_layer['config']",keras_layer['config'])
```
Comment (etfortin) [inline hls4ml/converters/keras_to_hls.py]: remove debug comment

**Pull request closed without merging**